effect/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -0,0 +1,1058 @@
+
+    /**
+     * The main interpreter loop for `IO` actions. For purely synchronous actions,
+     * this will run to completion unless required to yield to other fibers.
+     * For mixed actions, the loop will proceed no further than the first
+     * asynchronous boundary.
+     *
+     * @param io0 The `IO` to evaluate on the fiber.
+     */
+    final def evalSync(io0: IO[_]): Unit = {
+      try {
+        val maxopcount = YieldMaxOpCount
+
+        // Local variables. Do NOT accidentally capture any of these in a closure,
+        // or Scala will wrap them in ObjectRef and performance will plummet.
+        var result  : Try[Any]  = null
+        var curIo   : IO[Any]   = io0.asInstanceOf[IO[Any]]
+        var eval    : Boolean   = true
+        var opcount : Int       = 0
+
+        while (eval) {
+          (curIo.tag : @switch) match {
+            case IO.Tags.FlatMap =>
+              val io = curIo.asInstanceOf[IO.FlatMap[Any, Any]]
+
+              val nested = io.io
+
+              // A mini interpreter for the left side of FlatMap that evaluates
+              // anything that is 1-hop away. This eliminates heap usage for the
+              // happy path.
+              (nested.tag : @switch) match {
+                case IO.Tags.Point =>
+                  val io2 = nested.asInstanceOf[IO.Point[Any]]
+
+                  curIo = io.flatMapper(io2.value())
+
+                case IO.Tags.Strict =>
+                  val io2 = nested.asInstanceOf[IO.Strict[Any]]
+
+                  curIo = io.flatMapper(io2.value)
+
+                case IO.Tags.SyncEffect =>
+                  val io2 = nested.asInstanceOf[IO.SyncEffect[Any]]
+
+                  try {
+                    curIo = io.flatMapper(io2.effect())
+                  } catch {
+                    case t : Throwable if (nonFatal(t)) =>
+                      curIo = IO.Fail(t)
+                  }
+
+                case _ =>
+                  // Fallback case. We couldn't evaluate the LHS so we have to
+                  // use the stack:
+                  curIo = nested
+
+                  stack.push(io.flatMapper)
+              }
+
+            case IO.Tags.Point =>
+              val io = curIo.asInstanceOf[IO.Point[Any]]
+
+              val value = io.value()
+
+              curIo = nextInstr(value, stack)
+
+              if (curIo == null) {
+                eval   = false
+                result = \/-(value)
+              }
+
+            case IO.Tags.Strict =>
+              val io = curIo.asInstanceOf[IO.Strict[Any]]
+
+              val value = io.value
+
+              curIo = nextInstr(value, stack)
+
+              if (curIo == null) {
+                eval   = false
+                result = \/-(value)
+              }
+
+            case IO.Tags.SyncEffect =>
+              val io = curIo.asInstanceOf[IO.SyncEffect[Any]]
+
+              try {
+                val value = io.effect()
+
+                curIo = nextInstr(io.effect(), stack)
```

Comment:
Reevaluation of effect, should be `nextInstr(value, stack)`.

It might be a good idea to add a test which composes a somewhat complicated IO, which increments a global counter on each evaluation and tests that it ends up being equal to sum of effectively evaluated effects.